components/script/dom/htmlstyleelement.rs

```diff
@@ -229,11 +246,26 @@ impl VirtualMethods for HTMLStyleElement {
         }
 
         if context.tree_connected {
-            if let Some(s) = self.stylesheet.borrow_mut().take() {
-                self.clean_stylesheet_ownership();
-                self.stylesheet_list_owner()
-                    .remove_stylesheet(self.upcast(), &s)
+            self.remove_stylesheet();
+        }
+    }
+
+    fn attribute_mutated(&self, attr: &Attr, mutation: AttributeMutation) {
+        if let Some(s) = self.super_type() {
+            s.attribute_mutated(attr, mutation);
+        }
+
+        if attr.name() == "type" &&
+            self.upcast::<Node>().is_in_a_document_tree() &&
```

Comment:
I think `.is_in_a_document_tree()` should be 
`.is_in_a_document_tree() || is_in_a_shadow_tree()`, otherwise modifying the type attribute of a style element in a shadow tree won't cause the css to re-parse.